lib/posix: correct the meaning of CONFIG_MAX_PTHREAD_COUNT

wentongwu opened the pull request:
CONFIG_MAX_PTHREAD_COUNT seems to be implemented as the maximum number
of POSIX threads that can ever be created, rather than the maximum
number of active POSIX threads. Create an array to mark usage of
posix_thread_pool. On thread creation, use this array to find free pool
element and on thread exit mark posix_thread_pool element as free.

Fixes #15516.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

pfalcon reviewed (changes requested):
I'm not sure that the description here is clear enough, I'd suggest to try to formulate it a bit more clear.

Likewise, I'm not sure the the fix here is the best one:

1. Even from a quick look, it seems that the problem is that the code contains `pthread_num++` statements, but clearly missed `pthread_num--` at some right place.
2. Even if you like to deal with an array, there's already an array `posix_thread_pool`, so unlikely memory should be spent on adding another one.

I may be wrong here, please discuss what alternative ways of implementing a fix you considered and why they were rejected in preference of this patch.

andrewboie commented on lib/posix/pthread.c:
```diff
@@ -35,7 +35,8 @@ static const pthread_attr_t init_pthread_attrs = {
 
 static struct posix_thread posix_thread_pool[CONFIG_MAX_PTHREAD_COUNT];
-static u32_t pthread_num;
+static int posix_thread_pool_usage[CONFIG_MAX_PTHREAD_COUNT];
```
Can we use a bitfield instead and save 32x memory?

see thread_idx_alloc() and thread_idx_free() in kernel/userspace.c for an example

wentongwu commented:
> I'm not sure that the description here is clear enough, I'd suggest to try to formulate it a bit more clear.
> 
> Likewise, I'm not sure the the fix here is the best one:
> 
>     1. Even from a quick look, it seems that the problem is that the code contains `pthread_num++` statements, but clearly missed `pthread_num--` at some right place.
> 
>     2. Even if you like to deal with an array, there's already an array `posix_thread_pool`, so unlikely memory should be spent on adding another one.
> 
> 
> I may be wrong here, please discuss what alternative ways of implementing a fix you considered and why they were rejected in preference of this patch.
@pfalcon 
we should use array to identify which threads are used and which threads are available in pthread pool, pthread_num is not enough to do that. I will use bitfield to do that as @andrewboie mentioned and describe this change more clearly.

pfalcon commented:
> I will use bitfield to do that

Please use existing `posix_thread_pool[]`, or explain why it can't be used.

wentongwu commented:
> > I will use bit-field to do that
> 
> Please use existing `posix_thread_pool[]`, or explain why it can't be used.

posix_thread_pool[] is used to store struct posix_thread, yes I use it. I use another array to mark whether posix_thread_pool[i] is usable or not in the current commit. Using bit-field just to save memory.

pfalcon commented:
> posix_thread_pool[] is used to store struct posix_thread, yes I use it. I use another array

For the avoidance of doubt: the suggestion is not to use another array.

> I use another array to mark whether posix_thread_pool[i] is usable or not in the current commit.

There's no need to use another array, because `posix_thread_pool` itself contain enough information for that.

> Using bit-field just to save memory.

Sorry, but introducing more data structures is not a way to save memory, it's a way to spend it.

A mere look at the definition of `struct posix_thread` should show multiple ways to achieve the needed effect. An obvious one would be to use `enum pthread_state state;` field. However, `enum pthread_state` is defined subideally, and using that field would require an bit of extra legwork with initialization of `posix_thread_pool`. Again, instead of doing that, it would be better to fix `enum pthread_state` itself, e.g. to:

~~~
enum pthread_state {
        /* The thread structure is unallocated and available for (re)use. */
        PTHREAD_FREE = 0,
        /* The thread is running and joinable. */
        PTHREAD_JOINABLE,
        /* The thread is running and detached. */
        PTHREAD_DETACHED,
        /* A joinable thread exited and its return code is available. */
        PTHREAD_EXITED,
};
~~~

wentongwu commented:
> > posix_thread_pool[] is used to store struct posix_thread, yes I use it. I use another array
> 
> For the avoidance of doubt: the suggestion is not to use another array.
> 
> > I use another array to mark whether posix_thread_pool[i] is usable or not in the current commit.
> 
> There's no need to use another array, because `posix_thread_pool` itself contain enough information for that.
> 
> > Using bit-field just to save memory.
> 
> Sorry, but introducing more data structures is not a way to save memory, it's a way to spend it.
> 
> A mere look at the definition of `struct posix_thread` should show multiple ways to achieve the needed effect. An obvious one would be to use `enum pthread_state state;` field. However, `enum pthread_state` is defined subideally, and using that field would require a bit of extra legwork with initialization of `posix_thread_pool`. Again, instead of doing that, it would be better to fix `enum pthread_state` itself, e.g. to:
> 
> ```
> enum pthread_state {
>         /* The thread structure is unallocated and available for (re)use. */
>         PTHREAD_FREE = 0,
>         /* The thread is running and joinable. */
>         PTHREAD_JOINABLE,
>         /* The thread is running and detached. */
>         PTHREAD_DETACHED,
>         /* A joinable thread exited and its return code is available. */
>         PTHREAD_EXITED,
> };
> ```
@pfalcon I agree your idea, and I will review code more to make sure the pthread state is handled correctly, Thanks for your review.

pfalcon commented:
Thanks!

wentongwu commented:
@pfalcon could you please take a look the new changes?

pfalcon commented:
@wentongwu, I'm on vacation until Thurs, will look then. Thanks.

pfalcon commented on lib/posix/pthread.c:
```diff
@@ -145,6 +146,17 @@ int pthread_create(pthread_t *newthread, const pthread_attr_t *attr,
+			break;
+		}
+	}
+	pthread_num = i;
```
I'm sorry for nitpicking, but why do we need both of "i" and "pthread_num", if you end up unconditionally assigning one to another? Just use one var in the first place, appropriately named ("i" or "pthread_no" are examples of good names IMHO).

pfalcon commented on include/posix/pthread.h:
```diff
@@ -49,8 +49,8 @@ struct posix_thread {
 /* Pthread detach/joinable */
-#define PTHREAD_CREATE_JOINABLE     0
-#define PTHREAD_CREATE_DETACHED     1
+#define PTHREAD_CREATE_JOINABLE     PTHREAD_JOINABLE
```
This looks like a change unrelated to the purpose of the commit (as stated in the commit message)? Is it really needed? If not, I'd recommend to remove it. (And if it's really needed, the proper way would be to do it in a separate commit, with its message explaining why the change is done.)

wentongwu replied:
they will be used to init the state of struct pthread_state, so the value of PTHREAD_CREATE_JOINABLE and PTHREAD_CREATE_DETACHED should follow pthread_state enum.

pfalcon replied:
> they will be used to init the state of struct pthread_state

Sounds good, thanks.

pfalcon reviewed (changes requested):
Thanks, look how nice and simple it becomes! But I think we could clean up a bit more, can you please have a look?

pfalcon commented:
Thanks for updates, @wentongwu. Note that ideally, a discovered issue like this would be accompanied by a test to make sure that it now behaves correctly, and stays that way going forward. A test of course can come later ;-).

pfalcon commented:
@andrewboie, @nashif, please merge.

nashif reviewed (commented):
You probably also want to change the kconfig text and help to clarify this. The Kconfig is not clear at all.

pfalcon commented:
> You probably also want to change the kconfig text and help to clarify this. The Kconfig is not clear at all.

My suggestion: description ("Maximum pthread count in POSIX application") is good as it is, but replace current help text of:

> Mention maximum number of threads in POSIX compliant application.

with:

> Maximum number of simultaneously active threads in a POSIX application.

pfalcon commented:
@andrewboie, @nashif, I'd suggest to merge this as is. There're issues with help texts for a number of POSIX options, not just CONFIG_MAX_PTHREAD_COUNT, I'll submit a patch to fix them separately.

nashif reviewed (changes requested):
@wentongwu please change the help message

wentongwu commented:
@pfalcon done the update for the help message, please help review it. Thanks.

pfalcon commented:
I personally think that the change of the help message as done in this patch while solves one issue adds another level of confusion with those strange "in Zephyr" references, I suggested which help message should be used here: https://github.com/zephyrproject-rtos/zephyr/pull/15618#issuecomment-490021765 .

That said, I don't want to insist, and I'm sure that @nashif or @andrewboie will merge it anyway.

wentongwu committed:
lib/posix: correct the meaning of CONFIG_MAX_PTHREAD_COUNT

Current code implement CONFIG_MAX_PTHREAD_COUNT as the maximum number
of POSIX threads that can ever be created, rather than the maximum
number of active POSIX threads. Use pthread_state of struct posix_thread
to track the state of posix thread in posix_thread_pool so that we can
reuse the unused posix thread.

Fixes #15516.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>
